Add developer guide for 3D interactivity

kevinyamauchi opened the pull request:
This PR adds a guide for adding 3D interactivity via mouse event callbacks.

kevinyamauchi committed:
initial 3d interactivity guide

kevinyamauchi commented:
@alisterburt , here's the PR with the guide for adding 3D interactivity. I will flesh it out a bit more and add some figures tomorrow. Feel free to add your mouse drag stuff!

kevinyamauchi committed:
add figures

kevinyamauchi commented:
I have added the figures (attached below for convenience).

![3d_interaction_coordianates](https://user-images.githubusercontent.com/1120672/129695617-2128a4de-6457-4fff-928a-3510a23b1ef2.png)
![3d_interaction_click_line](https://user-images.githubusercontent.com/1120672/129695630-dbd07262-006c-41c7-b1bc-fb5d6f04dc8d.png)
![3d_interaction_ray_intersection](https://user-images.githubusercontent.com/1120672/129695640-b5b96fd0-032f-41d5-ac0b-53f365f43721.png)

kevinyamauchi committed:
add get_value explanation

kevinyamauchi commented:
@alisterburt , I think this is ready for a first round of review. I think there is more content we can add as we flesh out the 3D interactivity more, but I think this is a good start.

haesleinhuepf commented on guides/stable/3D_interactivity.md:
```diff
@@ -0,0 +1,131 @@
+![coordinate-systems](images/3d_interaction_coordianates.png)
+
+## In 3D mode, clicks are lines
+Since the 3D scene is rendered on a 2D surface (your screen), your mouse click does not map to a specific point in space. Thus, when rendering a scene in 3D, napari interprets your click as a line in the direction of the camera view that goes through the point that view ray intersects the camera plane.
```
```suggestion
Since the 3D scene is rendered on a 2D surface (your screen), your mouse click does not map to a specific point in space. As the view is a [parallel projection](https://en.wikipedia.org/wiki/Parallel_projection), napari can determine a line through 3D space that intersects the canvas where the user clicked.

alisterburt replied:
small note, technically we also support projection with perspective (shift, right click and drag I think?  `Camera.perspective`)) but this is a bit broken, doesn't work with 3D interactivity or the points layer in general...

kevinyamauchi replied:
Good point. I'll add a note about that.

haesleinhuepf commented on guides/stable/3D_interactivity.md:
```diff
@@ -0,0 +1,131 @@
+- dims_displayed: a list of the dimensions currently being displayed
+    in the viewer. This comes from viewer.dims.displayed.
+- dims_point: the indices for the data in view in world coordinates.
+
```
Just wondering, that means the following would roughly work, right?
```
def my_mouse_event(event)
    print(event.position)
```
Would it make sense to provide such a code snipped as example?

haesleinhuepf replied:
Ok, I just saw the example code further down :-)

haesleinhuepf commented on guides/stable/3D_interactivity.md:
```diff
@@ -0,0 +1,131 @@
+When a user clicks or moves the mouse in the canvas, napari emits a mouse event with the following properties:
+
+
+- pos: the position of the click in canvas coordinates.
```
I vote for a name that explains the thing. If a comment is necessary to explain what a variable contains, the variable has a bad name.
```suggestion
- canvas_position: the position of the click in canvas coordinates.
```

alisterburt replied:
Agree - we should refactor the events. In this case this is standard napari stuff, not directly related to 3D interactivity, so I think we should push forward, document the way things currently work and follow up with a general events refactor/documentation update.

Refactoring this will be quite a large API change for the events system in general which people definitely use quite a lot, we should probably do this as part of a `0.5.0` transition :)

haesleinhuepf commented on guides/stable/3D_interactivity.md:
```diff
@@ -0,0 +1,131 @@
+
+
+- pos: the position of the click in canvas coordinates.
+- position: the position of the click in world coordinates.
```
```suggestion
- world_position: the position of the click in world coordinates.
```
What does this variable contain? Above you mentioned that this cannot be a point. What is it then?

alisterburt replied:
it contains a point on the canvas plane (i.e. plane perpendicular to the ray being cast in parallel projection), transformed into the world coordinate system - agree this should be clearer and that it's non obvious. Maybe we should separate out these concepts generally @kevinyamauchi ? `camera_plane_point_world`

kevinyamauchi replied:
I will clarify the description of `pos` and `position` in the docs. I agree that both the names and descriptions are unclear. As @alisterburt  said, the names are pre-existing napari public API, so I think we should discuss in a separate issue.

kevinyamauchi replied:
I have attempted to clarify the meaning of the mouse events. They now read:

- **pos**: the position of the click in canvas coordinates.
- **position**: the position of the click in world coordinates. The point is located at the intersection of the click line (`view_direction`) and a plane parallel to the camera plane (i.e,. a plane normal to `view_direction`).
- **view_direction**: a unit vector giving the direction of the camera in
    world coordinates.
- **dims_displayed**: a list of the dimensions currently being displayed
    in the viewer. This comes from `viewer.dims.displayed`.
- **dims_point**: the indices for the data in view in world coordinates. This comes from `viewer.dims.point`.

kevinyamauchi replied:
I have made an issue on the napari repo to consider renaming the event properties.

https://github.com/napari/napari/issues/3189

haesleinhuepf commented on guides/stable/3D_interactivity.md:
```diff
@@ -0,0 +1,131 @@
+
+
+## Determining where the click intersects the data
+Each napari layer has a method called `get_ray_intersections()` that will return the points on the data bounding box that a given line will intersect (`start_point ` and `end_point `). `start_point` and `end_point` are  the end points of line segment that intersects the layer's axis-alinged data bounding box. `near_point` is the end point that is closest to the camera (i.e, the "first" intersecti ...
```
```suggestion
Each napari layer has a method called `get_ray_intersections()` that will return the points on the data bounding box that a given line will intersect (`start_point ` and `end_point `). `start_point` and `end_point` are the end points of a line segment that intersects the layer's axis-alinged data bounding box. `near_point` is the end point that is closest to the camera (i.e, the "first" intersection") and `far_point` is the end point that is farthest from the camera (i.e., the "last" intersection). You can use the line segment between `start_point` and `end_point` to interrogate the layer data that is "under" your cursor.
```

alisterburt replied:
```suggestion
Each napari layer has a method called `get_ray_intersections()` that will return the points on the data bounding box that a given line will intersect (`start_point ` and `end_point `). `start_point` and `end_point` are the end points of a line segment that intersects the layer's axis-alinged data bounding box. `near_point` is the end point that is closest to the camera (i.e, the "first" intersection) and `far_point` is the end point that is farthest from the camera (i.e., the "last" intersection). You can use the line segment between `start_point` and `end_point` to interrogate the layer data that is "under" your cursor.
```

small modification of @haesleinhuepf 's suggestion

haesleinhuepf commented on guides/stable/3D_interactivity.md:
````diff
@@ -0,0 +1,131 @@
+    	# use start_point and end_point to interrogate layer data
+```
+
+For an example implementation, see the [`cursor_ray.py`](https://github.com/napari/napari/blob/master/examples/cursor_ray.py) example.
````
```suggestion
For an example implementation, see [`cursor_ray.py`](https://github.com/napari/napari/blob/master/examples/cursor_ray.py).
```

haesleinhuepf commented on guides/stable/3D_interactivity.md:
```diff
@@ -0,0 +1,131 @@
+For an example implementation, see the [`cursor_ray.py`](https://github.com/napari/napari/blob/master/examples/cursor_ray.py) example.
+
+## Getting the layer data under the cursor
+We are currently implementing convenience methods to the layers (`layer.get_value()`) to get the layer data value underneath the cursor that is "on top" (i.e., closest to `start_point`). Like `layer.get_ray_intersections()`, `layer.get_value()` takes the click position, view direction, dims_displayed in either world or layer coordinates (see `world` argument) as input. Thus, it can be easily inte ...
```
```suggestion
There are convenience methods in the layer objects (`layer.get_value()`) to get the layer data value underneath the cursor that is "on top" (i.e., closest to `start_point`). Like `layer.get_ray_intersections()`, `layer.get_value()` takes the click position, view direction, dims_displayed in either world or layer coordinates (see `world` argument) as input. Thus, it can be easily integrated into a mouse event callback. Note that `layer.get_value()` returns `None` if the layer is not currently visible. See the docstring below for details.
```

alisterburt replied:
same point as above, should we provide the 'arbitrary' interrogation as a util function somewhere and have users pass the mouse event directly to the method on the layer?

kevinyamauchi replied:
Yeah, maybe we add something like `get_value_from_mouse_event()`? Too long?

kevinyamauchi replied:
As with the`get_ray_intersections` function that takes an event, I think the `get_value` one should also be on the layer so that the developer doesn't need to import anything extra to make a mouse callback.

haesleinhuepf commented on guides/stable/3D_interactivity.md:
````diff
@@ -0,0 +1,131 @@
+            Value of the data. If the layer is not visible return None.
+```
+
+We have implemented `layer.get_value()` for the layers below. All other layers with return `None` from `layer.get_value()` when called while the viewer is in a 3D rendering mode.
````
```suggestion
We have implemented `layer.get_value()` for the layers below. All other layers with return `None` from `layer.get_value()` when called while the viewer is in a 3D rendering mode.
```
This sounds like not ideal. Maybe create an issue about what's happening here on development side at the moment and link the issue...

alisterburt replied:
issue added at napari/napari#3187

alisterburt replied:
```suggestion
Progress on the implementation of `layer.get_value()` can be found at [#3187](https://github.com/napari/napari/issues/3187). Layers for which this is not yet implemented will return `None` if `layer.get_value()` is called whilst the viewer is in a 3D rendering mode.
```

haesleinhuepf reviewed (changes requested):
Very well described @kevinyamauchi ! I would just suggest a minor API change: The variable names `pos` and `position` can be improved, e.g. "canvas_position" and "world_position". 

I also suggested some minor text changes.

👍

alisterburt commented on guides/stable/3D_interactivity.md:
```diff
@@ -0,0 +1,131 @@
+# 3D interactivity
+
+## Coordinate systems in napari
+In napari, there are three main coordinate systems: (1) canvas, (2) world, and (3) layer. The canvas coordinates system is the 2D coordinate system of the canvas on which the scene is rendered. World coordinates are the nD coordinates of the entire scene. As the name suggests, the layer coordinates is the nD coordinate system of the data in a given layer. Layer coordinates are specific the each l ...
```
```suggestion
In napari, there are three main coordinate systems: (1) canvas, (2) world, and (3) layer. The canvas coordinates system is the 2D coordinate system of the canvas on which the scene is rendered. World coordinates are the nD coordinates of the entire scene. As the name suggests, layer coordinates are the nD coordinate system of the data in a given layer. Layer coordinates are specific the each layer's data and are related to the world coordinate system via the layer transforms.
```

alisterburt commented on guides/stable/3D_interactivity.md:
````diff
@@ -0,0 +1,131 @@
+See the `get_ray_intersection()` docstrings below for details. Note that if the line does not intersect the data bounding box (i.e., the click was outside of the data), `start_point` and `end_point` are `None`. The `position` and `view_direction` should be provided as world coordinates if `world` is set to True and in layer coordinates if `world` is set to `False`.
+
+```
+def get_ray_intersections(
+    self,
+    position: List[float],
+    view_direction: np.ndarray,
+    dims_displayed: List[int],
+    world: bool = True,
+) -> Union[Tuple[np.ndarray, np.ndarray], Tuple[None, None]]:
````
reading this makes me think - does it make sense to simplify this interface a bit for the user. We could hide a lot of the complexity in a utility function which does what this method does and have this method take only the mouse event?

`Layer.get_ray_intersections(event)` -> start and end point

kevinyamauchi replied:
I am open to adding a convenience method or function that takes the mous event. I think it is important to keep a public function that takes the keyword arguments above. That way if a user wants to interrogate their data outside of a mouse event callback, they don't have to create a mock `Event`.

alisterburt replied:
cool I'll try to get this done later today. Are you okay with having that function be public but tucked away in `napari.layers.utils` or would you prefer to have both on the layer? I think I would prefer tucking it away and pointing to it from these docs

kevinyamauchi replied:
I think it would be nice to have the one that takes the event also on the layer. That way the user doesn't have to import anything to create a mouse event callback.

alisterburt replied:
slight misunderstanding I think, what I'm proposing is
- more complex utility function in some utils file (but public)
- simpler function which takes a mouse event directly on the layer

kevinyamauchi replied:
Got it. Is there a reason not to have both on the layer? The function that takes the event basically becomes a passthrough. I think we'd want to leave at least `layer.get_value()` as is since that has been part of the public API for a while.

alisterburt replied:
no specific reason - happy for both to be on the layer with the `_from_mouse_event` nomenclature!

kevinyamauchi replied:
Cool, let's have them both on the layer then. Thanks!

alisterburt commented on guides/stable/3D_interactivity.md:
````diff
@@ -0,0 +1,131 @@
+```
+
+## Adding 3D interactivity via mouse events
+Custom 3D interactivity can be added via mouse callbacks. The `layer.get_ray_intersections()` function has been designed to work seamlessly with the napari mouse callback event. You can pass the mouse callback event properties to `layer.get_ray_intersections()` get the `start_point` and `end_point` of where the click line intersects the layer data. 
````
see above about seamlessness

kevinyamauchi replied:
We can update here when you've added the convenience function!

alisterburt commented on guides/stable/3D_interactivity.md:
````diff
@@ -0,0 +1,131 @@
+```
+
+We have implemented `layer.get_value()` for the layers below. All other layers with return `None` from `layer.get_value()` when called while the viewer is in a 3D rendering mode.
+- `Labels`
````
```suggestion
```
remove wrt change above

alisterburt reviewed (commented):
super nice man! Took a first pass and this is a really clear explanation for the dev side - love the figures too. 

I agree with everything @haesleinhuepf said and left a few comments below!

What do you think about layer coordinates vs. data coordinates as a naming thing? I prefer and use data in my head, don't feel too strongly about it but the transforms refer to data (`data2world` etc) so might be worth sticking to that

tlambert03 commented:
lovely!  great figures :)

kevinyamauchi committed:
Apply suggestions from @alisterburt and @haesleinhuepf

Co-authored-by: alisterburt <alisterburt@gmail.com>
Co-authored-by: Robert Haase <haesleinhuepf@users.noreply.github.com>

kevinyamauchi committed:
clarify mouse event properties

kevinyamauchi commented:
> lovely! great figures :)

Thanks, @tlambert03 !

Thank you for the reviews, @haesleinhuepf and @alisterburt ! I believe I have addressed your comments.

@alisterburt , perhaps if you think this is good to go, we can merge this and then you can add the dragging, etc. stuff in a follow-up PR?

alisterburt commented:
sounds good to me @kevinyamauchi - let's merge!

sofroniewn commented:
Guide looks great! @kne42 @codemonkey800 same question though that I had in https://github.com/napari/napari.github.io/pull/145#issuecomment-900464352, are we going to overwrite this next time we build our docs? Did this need to get contributed to the main napari repo instead? Apologies that I've lost track of all this, I just want to make sure it all works out as we expect!!

kne42 commented:
Yes, this will be overwritten since it needs to be contributed to the main napari repo.

![image](https://user-images.githubusercontent.com/29165011/129948276-6290da05-83b5-4e22-b98f-d074e0cd51bb.png)

Any "copied source" listed in the above map needs to be modified in the main napari repo. Note that the "Learn" section has since been renamed to "Developer Guides".

kevinyamauchi commented:
Ah oops, sorry about that. Thanks for the explanation, @kne42 ! 

I'll make a PR to the napari repo tomorrow.
